boards/nucleo-l1: use LSI as default RTC clock

haukepetersen opened the pull request:
fixes #6502
~~rebased on #6499 (needed for testing)~~

seems like not all `nucleo-l1` boards have an external low-speed oscillator (LSE) attached. So I propose to make it configurable in the board's periph_conf and choose the LSI (internal low-speed oscillator) per default.

vincent-d commented:
The problem of the LSI is its accuracy. But it works if application does not require accuracy for long periods I guess.

However, I took a new look to the RTC application note. It seems that the stm32l1 LSI is a 37kHz oscillator, the the prescaler register needs to be adapted. (AN3371, part 1.1.2, page 9)

PeterKietzmann commented:
I'm on current master (includes #6499) **without** this PR running tests/periph_rtc. The output looks as expected:
```
2017-01-27 16:53:49,806 - INFO # RIOT RTC low-level driver test
2017-01-27 16:53:49,811 - INFO # This test will display 'Alarm' in 10 seconds
2017-01-27 16:53:49,811 - INFO # 
2017-01-27 16:53:49,812 - INFO # Initializing the RTC driver
2017-01-27 16:53:49,817 - INFO # Setting clock to 2011-12-13 14:15:15
2017-01-27 16:53:49,823 - INFO # Clock set to 2011-12-13 14:15:15
2017-01-27 16:53:49,824 - INFO # Setting alarm to 2011-12-13 14:15:25
2017-01-27 16:53:49,829 - INFO # Alarm set to 2011-12-13 14:15:25
2017-01-27 16:53:49,830 - INFO # The alarm should trigger every 10 seconds for 4 times.
2017-01-27 16:53:59,825 - INFO # Alarm!
2017-01-27 16:54:09,821 - INFO # Alarm!
2017-01-27 16:54:19,823 - INFO # Alarm!
2017-01-27 16:54:29,825 - INFO # Alarm!
```
So does **with** this PR.. Do we actually need it?

pwillem commented on cpu/stm32_common/periph/rtc.c:
```diff
@@ -221,22 +221,29 @@ void rtc_clear_alarm(void)
+#else
+    /* Enable the LSI clock */
+    RCC->CSR = RCC_CSR_LSION;
+    while (!(RCC->CSR & RCC_CSR_LSIRDY)) {}
```
For a **F0** this didn't work. The LSI is off after each reboot but [(RCC->BDCR & RCC_BDCR_RTCEN) == 1](https://github.com/RIOT-OS/RIOT/pull/6504/files#diff-bdada3d6cf063f73ecca7e6bd807935dR67) so rtc_poweron() isn't called. Adding these 2 lines to clk_init() in cpu.c solved it for me

haukepetersen replied:
will look into this.

haukepetersen commented:
@PeterKietzmann: probably your board is equipped with an external low-speed oscillator, so it works on master. But mine isn't, so it doesnt work on master -> ergo: this PR is needed.

vincent-d commented:
I proposed some more fixes in https://github.com/haukepetersen/RIOT/pull/37
- prescaler values updated
- LSI for stm32fx

But I don't really like the LSI as default...

kYc0o commented:
Confirmed that it works for nucleo-l152, but it breaks on nucleo-f091. Should this bug be postponed?

PeterKietzmann commented:
I'd say yes. The release should go out tomorrow. Testing has mostly passed and I already added this issue to known issues in the preliminary (still invisible) release notes. Will take care of all 2017.01 labelled PRs tomorrow.

vincent-d commented:
@kYc0o weird, I tested on a nucleo-f091 here and it worked :/

Anyway, I agree it should be postponed.

kYc0o commented:
Well, mine gets stuck in setting the clock:

```console
2017-02-02 12:42:52,413 - INFO # main(): This is RIOT! (Version: 2017.04-devel-8-gd1df5-snake.lan-pr/stm32l_lsi)
2017-02-02 12:42:52,413 - INFO # 
2017-02-02 12:42:52,415 - INFO # RIOT RTC low-level driver test
2017-02-02 12:42:52,420 - INFO # This test will display 'Alarm' in 10 seconds
2017-02-02 12:42:52,420 - INFO # 
2017-02-02 12:42:52,421 - INFO # Initializing the RTC driver
2017-02-02 12:42:52,425 - INFO # Setting clock to 2011-12-13 14:15:15
```

Same question: Which revision of nucleo-f091 do you have? does it have an external oscillator?

vincent-d commented:
My nucleo-f091 has an external oscillator. I've just retested on it (with `#define CLOCK_HAS_LSE       (0) `) and I got the same issue, it gets stuck.

On a nucleo-f042 without LSE, it works fine.

vincent-d commented:
This is not a RTC issue, this is a timer configuration issue. It get stuck in `xtimer_usleep`.

vincent-d commented:
Seems related to #6494

kYc0o commented:
Oh yeah! We need to merge first #6494 and then this will work for both boards. I'll add the Waiting for other PR flag.

I rebased this one with #6494 and everything works for both nucleo-l1 and nucleo-f091 boards.

kYc0o commented:
@haukepetersen can you squash and optionally rebase to include #6494 ? Then I think this is ready.

PeterKietzmann commented:
Ping @haukepetersen

aabadie commented:
any update on this one @haukepetersen ?

kYc0o commented:
ping @haukepetersen

haukepetersen commented:
sorry for the delay. Rebased on master and thought about the problem a little bit (while working on #6814)...

I think I changed my mind and I'd like to propose a fix in another direction:
- using LSE (over LSI) per default seems valid to me
- BUT: we should make it easy to switch between them -> let's try to unify the LSE/LSI config for all STM32 CPUs
- let's remove the LSE/LSI configuration from the RTC driver and move it into the clock init code in `cpu.c` for all STM32 CPUs
- I further propose to always initialize and run the low speed clock, it (a) does not use up much energy anyhow, it (b) is used in almost all low power applications, and it (c) makes the RTC code simpler. Also the low speed clock might be used by other peripherals than the RTC, so controlling this in the RTC driver will cause trouble...

vincent-d commented:
Sounds reasonable. We just need to make sure we reset the backup domain only if needed (i.e. the clock was not already initialized), otherwise we could stop counting and with boards with backup power provided it could be an issue.

haukepetersen commented:
Yes, that makes sense. One more reason to handle the clock LS(I/E) clock only at one place in the code.

haukepetersen commented:
I am looking into refining the RTT/RTC interfaces right now, and will in the same run look into a way to handle this for all STM32 CPUs

kYc0o commented:
Can we merge this one for this release? I think that as long as the bug is solved, we can wait for the rework on this interface and merge it for the next release.

AFAIK this PR is a working state and can be merged asap (besides the need of squash).

haukepetersen commented:
nope, won't merge this in the current state. Have a far better solution to the problem now (based on some clock initialization restructuring as done in #6907), so I will adapt this PR accordingly.

kYc0o commented:
And you think we can get it for the release? That's only my point, that is broken and it will stay broken for this release...

haukepetersen commented:
the solution as in this PR is sub-optimal, found something better by now: the `rtc` module should utilize `stmclk_enable_lfclk()` from newly introduced `stm32_common/include/stmclk.h` interface. The problem: this is not implement for many STM family members yet -> so once this is done, everything in this PR falls into place. Opened #7123 for tracking the implementation progress.

haukepetersen commented:
pushed an update to the rtc driver using the stmclk interface -> much easier to the eye... But as said: all WIP as of now...

haukepetersen commented:
closing in favor of #7504

kYc0o commented:
#8024 led me here... So, I propose to enable the internal oscillator only for the nucleo-l152 board, since anyways nobody will use it for accurate RTC and all stm32l152 have LSI. Should I open a PR for that small change? I tested it on a board with external oscillator and works correctly with LSI set.
